### ISSUE TITLE
Difficulty Reproducing HalfCheetah-v2 SAC Results

### ISSUE THREAD

**Issue by xanderdunn** [vitchyr/rlkit#128]
Huge thanks for providing this implementation, it's very high quality. 

I'm having difficulty reproducing the results of the original SAC paper using the provided examples/sac.py script.

The [paper](https://arxiv.org/pdf/1812.05905.pdf) reports a mean return of 15,000 in 3M steps (blue and orange lines are SAC):
<img width="505" alt="Screen Shot 2021-02-21 at 08 52 07" src="https://user-images.githubusercontent.com/1313618/108632125-23025e80-7422-11eb-9c40-c27028779628.png">

My runs on the unmodified examples/sac.py script appear to be considerably less sample efficient:
<img width="495" alt="Screen Shot 2021-02-21 at 08 58 18" src="https://user-images.githubusercontent.com/1313618/108632301-03b80100-7423-11eb-83d9-8accabfc3399.png">

My runs are pretty consistently achieving 13,000 average return on 10M steps. They may eventually get to 15,000 average return if left to run for millions of steps further, but my runs are requiring more than 3x the number of steps to achieve 13k vs 15k mean return. 

I have found that results can vary greatly from run to run. Notice the pink line in my above chart that does poorly. Is the paper doing many runs and reporting the best? I didn't see this mentioned in the Experiments section of the paper.

It appears to me that the hyper parameters shown in the paper are the same in the script:
<img width="1207" alt="Screen Shot 2021-02-21 at 08 50 22" src="https://user-images.githubusercontent.com/1313618/108632067-e0d91d00-7421-11eb-85f2-ae453ad5896a.png">

Am I interpreting the "num total steps" and "Returns Mean" correctly? Do you know what might cause this difference in sample efficiency and final return?

**Comment by vitchyr** [comment]
Hi, thanks for pointing this out. One possible cause for this difference is that this implementation alternates between sampling entire trajectories and taking gradient steps, where as the original SAC paper alternates between one environment step and one gradient step. It's hard to compare the two exactly, but I'm guessing that something small like increase `num_trains_per_train_loop` would compensate for this difference.

Another possible differences are differences in network initialization or very minor differences in the Adam optimizer implementation (I've seen people talk about this, though I don't particularly suspect this).

**Comment by xanderdunn** [comment]
@vitchyr Thanks very much, I will try increasing `num_trains_per_train_loop`.

I don't see mention in the SAC paper of how the network's weights were initialized. I might look at the [official implementation](https://github.com/rail-berkeley/softlearning) to see if it differs.

**Comment by xanderdunn** [comment]
What values of `num_trains_per_train_loop` would you recommend trying? With values 1000-3000 I'm not seeing a large difference in sample efficiency:

<img width="1716" alt="Screen Shot 2021-02-22 at 08 10 38" src="https://user-images.githubusercontent.com/1313618/108735470-74bfed00-74e5-11eb-9ce9-a21905c4cb99.png">

Light blue is the default 1000 and the others are 2000 or 3000. The best I'm seeing by step 3M is mean return 10.2k, vs. the paper's 15k.

**Comment by vitchyr** [comment]
Thanks for trying that. My main suspicion then is the difference between the batch data collection versus the intertwining data collection that could cause the difference. If you want to investigate this, replace the [evaluation path collector](https://github.com/vitchyr/rlkit/blob/7daf34b0ef2277d545a0ee792399a2ae6c3fb6ad/examples/sac.py#L51) with a [step collector](https://github.com/vitchyr/rlkit/blob/7daf34b0ef2277d545a0ee792399a2ae6c3fb6ad/rlkit/samplers/data_collector/step_collector.py#L10) and replace the [batch RL algorithm](https://github.com/vitchyr/rlkit/blob/7daf34b0ef2277d545a0ee792399a2ae6c3fb6ad/examples/sac.py#L68) with an [online RL algorithm](https://github.com/vitchyr/rlkit/blob/7daf34b0ef2277d545a0ee792399a2ae6c3fb6ad/rlkit/torch/torch_rl_algorithm.py#L13). It might take a few more edits to get it to run, but these components should be fairly plug-and-play.

**Comment by xanderdunn** [comment]
Thanks again for your help @vitchyr. 

It looks like this issue in the soft learning repo is related: rail-berkeley/softlearning#75

However, I managed to get the same experiment running in soft learning and found the results matched those in the paper.  Running this:
```
softlearning run_example_local examples.development \
    --algorithm SAC \
    --universe gym \
    --domain HalfCheetah \
    --task v2 \
    --exp-name my-sac-experiment-1 \
    --checkpoint-frequency 1000
```

I got these results on four different seeds:
<img width="1705" alt="Screen Shot 2021-02-23 at 07 50 32" src="https://user-images.githubusercontent.com/1313618/108869227-d8145280-75ab-11eb-8452-61cf67900cb4.png">

These results match the paper's reported result achieving ~15,000 mean return on the first 3M timesteps. The evaluation mean return was >15k for all runs. Note that each of these runs took 10.7 hours.

Compare to rlkit runs with four different values of num_trains_per_train_loop:
<img width="1707" alt="Screen Shot 2021-02-23 at 07 47 31" src="https://user-images.githubusercontent.com/1313618/108869451-127def80-75ac-11eb-9722-e2e80b56f075.png">

Mean return on the first 3M tilmesteps ranges from 6,200-11,000. Due to the high values of num_trains_per_train_loop, these results also took longer to compute. The best performing one, with `num_trains_per_train_loop==5000`, took 14 hours under the same hardware conditions.

rlkit has more RL algorithms implemented and is better maintained, but for now I will continue with the tensorflow implementation since the baseline is immediately accessible. The sample and computational efficiency are important aspects for our work.

**Comment by jihwan-jeong** [comment]
Hi Vitchyr and Xander,

Although this may not be directly related to this issue, look at the figure below in which the average return kinda better matches the performance that Xander referred to in the above thread. 
![image](https://user-images.githubusercontent.com/51473118/124695048-6bc43b00-df1d-11eb-89d8-ff4f96b10ace.png)

The result is based only on a single run, so I can't say for sure whether the issue is resolved. Others may double-check if they can..

Basically, I consulted the softlearning repo that Xander linked and found a discrepancy w.r.t. the rlkit implementation.
Specifically, the order of gradient updates is different: in rlkit, the actor update and the critic update are mixed in order, which actually resulted in errors in recent versions of PyTorch (>=1.5) (see [this discussion](https://discuss.pytorch.org/t/solved-pytorch1-5-runtimeerror-one-of-the-variables-needed-for-gradient-computation-has-been-modified-by-an-inplace-operation/90256/5) in PyTorch forum). In contrast, what the softlearning repo does is to update critic, actor, and alpha in sequence. To implement that, I just kinda switched the order of gradient updates in the code: 

```python
    def train_from_torch(self, batch):
        rewards = batch['rewards']
        terminals = batch['terminals']
        obs = batch['observations']
        actions = batch['actions']
        next_obs = batch['next_observations']
        """
        As in 'github.com/rail-berkeley/softlearning', parameters are trained in the order as follows:
            (1) Critic
            (2) Actor
            (3) Alpha
        Otherwise, in PyTorch (>=1.5), the inplace operation error occurs.  
        """

        # Firstly, get the current alpha value
        if self.use_automatic_entropy_tuning:
            alpha = self.log_alpha.exp().detach()
        else:
            alpha = 1.

        """
        Critic loss
        """
        # Make sure policy accounts for squashing functions like tanh correctly!
        new_next_actions, _, _, new_log_pi, *_ = self.policy(
            next_obs, reparameterize=True, return_log_prob=True,
        )

        target_q_values = torch.min(
            self.target_qf1(next_obs, new_next_actions),
            self.target_qf2(next_obs, new_next_actions),
        ) - alpha * new_log_pi

        q_target = self.reward_scale * rewards + (1. - terminals) * self.discount * target_q_values
        q_target.detach_()

        q1_pred = self.qf1(obs, actions)
        q2_pred = self.qf2(obs, actions)

        qf1_loss = self.qf_criterion(q1_pred, q_target)
        qf2_loss = self.qf_criterion(q2_pred, q_target)

        # Update critics
        self.qf1_optimizer.zero_grad()
        qf1_loss.backward()
        self.qf1_optimizer.step()

        self.qf2_optimizer.zero_grad()
        qf2_loss.backward()
        self.qf2_optimizer.step()

        """
        Policy loss (updating actor)
        """
        new_obs_actions, policy_mean, policy_log_std, log_pi, *_ = self.policy(
            obs, reparameterize=True, return_log_prob=True,
        )
        q_new_actions = torch.min(
            self.qf1(obs, new_obs_actions),
            self.qf2(obs, new_obs_actions),
        )
        policy_loss = (alpha * log_pi - q_new_actions).mean()

        # Update actor
        self.policy_optimizer.zero_grad()
        policy_loss.backward()
        self.policy_optimizer.step()

        """
        Alpha loss
        """
        if self.use_automatic_entropy_tuning:
            new_obs_actions, _, _, log_pis, *_ = self.policy(
                obs, reparameterize=True, return_log_prob=True,
            )
            alpha_loss = -(self.log_alpha * (log_pis + self.target_entropy).detach()).mean()
            self.alpha_optimizer.zero_grad()
            alpha_loss.backward()
            self.alpha_optimizer.step()
            alpha = self.log_alpha.exp().detach()

        """
        Soft Updates
        """
        if self._n_train_steps_total % self.target_update_period == 0:
            ptu.soft_update_from_to(
                self.qf1, self.target_qf1, self.soft_target_tau
            )
            ptu.soft_update_from_to(
                self.qf2, self.target_qf2, self.soft_target_tau
            )
```

Hope this helps in resolving this issue :)

**Comment by ZhenhuiTang** [comment]
Hii, where could I see the results, when I run "python3 examples/ddpg.py" ? I could not find the 'output' file.

**Comment by A373k** [comment]
how avg return vs step is caluted as avg return vs  episode count makes sense
